README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](LICENSE.txt)
[![Webex Events](https://github.com/SocioEvents/webex-events-java-sdk/actions/workflows/maven.yml/badge.svg)](https://github.com/SocioEvents/webex-events-java-sdk/actions)

⚠️ This library has not been released yet.
# Webex Events Api Java SDK

Webex Events provides a range of additional SDKs to accelerate your development process.
They allow a standardized way for developers to interact with and leverage the features and functionalities.
Pre-built code modules will help access the APIs with your private keys, simplifying data gathering and update flows.

Requirements
-----------------

Java 17+

Installation
-----------------
For maven:
```xml
<dependency>
    <groupId>com.webex.events</groupId>
    <artifactId>webex-events</artifactId>
    <version>desired version</version>
</dependency>
```
```shell
$ mvn clean install
```

For Gradle:
```
dependencies {
    compile "com.webex.events:webex-events:<version>"
}
```
```
$ gradle build --refresh-dependencies
```

Configuration
-----------------
```java
Configuration configuration = new Configuration();
configuration
        .setAccessToken("sk_live_your_access_token")
        .setTimeout((byte) 30)
        .setMaxRetries((byte) 5);
```

Usage
-----------------
```java
String query = "query CurrenciesList{ currenciesList{ isoCode }}";
String operationName = "CurrenciesList";
HashMap<String, Object> variables = new HashMap<>();
HashMap<String, Object> headers = new HashMap<>();

try {
    Response response = Client.query(query, operationName, variables, headers, configuration);
}catch (DailyQuotaIsReachedError e) {
    // Handle what will happen if daily quota is reached here.
}catch (SecondBasedQuotaIsReachedError e) {
    int retryAfter = e.response().getRateLimiter().getSecondlyRetryAfterInMs()
    if (retryAfter > 0) {
        Thread.sleep(retryAfter);
        // Retry the request.
    }
}catch (ConflictError e) {
    Thread.sleep(250);
    // Retry the request
}
```

By default some HTTP statuses are retriable such as `408, 409, 429, 502, 503, 504`. This library tries this status
codes 5 times by default. If this is not sufficient, increase max retry count through Configuration class or re-catch 
the exceptions to implement your logic here. 

For Introspection Query
-----------------
```java
String json = Client.doIntrospectQuery(configuration);
```

Idempotency
-----------------
The API supports idempotency for safely retrying requests without accidentally performing the same operation twice.
When doing a mutation request, use an idempotency key. If a connection error occurs, you can repeat
the request without risk of creating a second object or performing the update twice.

To perform mutation request, you must add a header which contains the idempotency key such as
`Idempotency-Key: <your key>`. The SDK does not produce an Idempotency Key on behalf of you if it is missed.
The SDK also validates the key on runtime, if it is not valid UUID token it will raise an exception. Here is an example
like the following:

```java
String query = "mutation TrackDelete($input: TrackDeleteInput!) {trackDelete(input: $input) {success}}";
String operationName = "TrackDelete";
HashMap<String, Object> variables = new HashMap<>();
HashMap<String, Object> headers = new HashMap<>();

HashMap<String, Object> input = new HashMap<>();
input.put("ids", new int[]{1,2,3});
input.put("eventId", 1);
variables.put("input", input);

headers.put("Idempotency-Key", UUID.randomUUID().toString());

Response response = Client.query(query, operationName, variables, headers, configuration);
```

Telemetry Data Collection
-----------------
Webex Events collects telemetry data, including hostname, operating system, language and SDK version, via API requests.
This information allows us to improve our services and track any usage-related faults/issues. We handle all data with
the utmost respect for your privacy. For more details, please refer to the Privacy Policy at https://www.cisco.com/c/en/us/about/legal/privacy-full.html

Development
-----------------

After checking out the repo, install dependencies. Then run tests.

Contributing
-----------------
Please see the [contributing guidelines](CONTRIBUTING.md).

License
-----------------

The library is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

Code of Conduct
-----------------

Everyone interacting in the Webex Events API project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/SocioEvents/webex-events-java-sdk/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2023-11-25

- Initial release


src/main/java/com/webex/events/Client.java
package com.webex.events;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.webex.events.exceptions.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.IntStream;

public class Client {
    final static Logger logger = LoggerFactory.getLogger(Client.class);
    public static final int[] RETRIABLE_HTTP_STATUSES = {408, 409, 429, 502, 503, 504};

    public static String doIntrospectQuery(Configuration configuration) throws Exception {
        logger.debug("Doing introspection query...");
        Response response = query(
                Helpers.getIntrospectionQuery(),
                "IntrospectionQuery",
                new HashMap<>(),
                new HashMap<>(),
                configuration
        );
        return response.toString();
    }

    public static Response query(
            String query,
            String operationName,
            HashMap<String, Object> variables,
            HashMap<String, Object> headers,
            Configuration config
    ) throws Exception {

        Helpers.validateIdempotencyKey(headers.get("Idempotency-Key"));
        Helpers.validateAccessTokenExistence(config.getAccessToken());

        HashMap<String, Object> params = new HashMap<>();
        params.put("query", query);
        params.put("operationName", operationName);
        params.put("variables", variables);

        ObjectMapper objectMapper = new ObjectMapper();
        String json = objectMapper.writeValueAsString(params);

        headers.put("Content-Type", "application/json");
        headers.put("Authorization", "Bearer " + config.getAccessToken());
        headers.put("X-Sdk-Name", "Java SDK");
        headers.put("X-Sdk-Version", Helpers.getSDKVersion());
        headers.put("X-Sdk-Lang-Version", System.getProperty("java.version"));
        headers.put("User-Agent", Helpers.getUserAgent());

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(Helpers.getUri(config.getAccessToken()));

        for (Map.Entry<String, Object> header : headers.entrySet()) {
            requestBuilder.header(header.getKey(), (String) header.getValue());
        }

        HttpRequest request = requestBuilder
                .timeout(Duration.ofSeconds(config.getTimeout()))
                .POST(HttpRequest.BodyPublishers.ofString(json))
                .build();

        long startTime = System.currentTimeMillis();
        HttpClient httpClient = HttpClient.newHttpClient();
        Response response = doOrRetryTheRequest(config, httpClient, request, operationName);
        long endTime = System.currentTimeMillis();

        response.setTimeSpentInMs((int) (endTime - startTime));
        response.setRequestBody(json);
        response.setRateLimiter(new RateLimiter(response));

        logger.info(
                "Executing {} query is finished with {} status code. It took {} ms and retried {} times.",
                operationName,
                response.status(),
                response.getTimeSpendInMs(),
                response.getRetryCount()
        );

        if (response.status() > 299) {
            logger.error("Executing {} query is failed. Received status code is {}", operationName, response.status());
            manageErrorState(response);
        }

        return response;
    }

    private static Response doOrRetryTheRequest(Configuration config, HttpClient httpClient, HttpRequest request, String operationName) {
        logger.info("Executing {} query for the first time to {}", operationName, Helpers.getUri(config.getAccessToken()));

        Response response;
        try {
            response = new Response(httpClient.send(request, HttpResponse.BodyHandlers.ofString()));
            buildErrorResponse(response);
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        logger.error("{} http status received for {} query.", response.status(), operationName);

        Response finalResponse = response;
        int i = 0;
        double waitInterval = 250.0;
        double waitRate = 1.4;
        if (IntStream.of(RETRIABLE_HTTP_STATUSES).anyMatch(x -> x == finalResponse.status())) {
            while ((i < config.getMaxRetries())) {
                i++;
                try {
                    response = new Response(httpClient.send(request, HttpResponse.BodyHandlers.ofString()));
                    buildErrorResponse(response);
                    ErrorResponse errorResponse = response.getErrorResponse();
                    if (errorResponse != null && errorResponse.dailyAvailableCostIsReached()) {
                        break;
                    }

                } catch (IOException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
                if (response.status() < 300) {
                    break;
                } else {
                    waitInterval *= waitRate;
                    logger.info(
                            "The HTTP request is being restarted for {} query. Waiting for {} ms...",
                            operationName,
                            (int) waitInterval
                    );

                    try {
                        Thread.sleep((int) waitInterval);
                    } catch (InterruptedException e) {
                        throw new RuntimeException(e);
                    }
                }
            }
        }
        response.setRetryCount(i);
        return response;
    }

    private static void buildErrorResponse(Response response) {
        try {
            if (response.status() > 299) {
                ObjectMapper mapper = new ObjectMapper();
                ErrorResponse errResponse;
                errResponse = mapper.readValue(response.body(), ErrorResponse.class);
                response.setErrorResponse(errResponse);
            }
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
    private static void manageErrorState(Response response) throws Exception {
        ErrorResponse errorResponse = response.getErrorResponse();
        switch (response.status()) {
            case 400:
                if (errorResponse.isInvalidToken()) {
                    throw new InvalidAccessTokenError(response);
                } else if (errorResponse.isTokenIsExpired()) {
                    throw new AccessTokenIsExpiredError(response);
                } else {
                    throw new BadRequestError(response);
                }
            case 401:
                throw new AuthenticationRequiredError(response);
            case 403:
                throw new AuthorizationFailedError(response);
            case 404:
                throw new ResourceNotFoundError(response);
            case 408:
                throw new RequestTimeoutError(response);
            case 409:
                throw new ConflictError(response);
            case 413:
                throw new QueryComplexityIsTooHighError(response);
            case 422:
                throw new UnprocessableEntityError(response);
            case 429:
                if (errorResponse.dailyAvailableCostIsReached()) {
                    throw new DailyQuotaIsReachedError(response);
                }

                if (errorResponse.availableCostIsReached()) {
                    throw new SecondBasedQuotaIsReachedError(response);
                }
            case 500:
                throw new ServerError(response);
            case 502:
                throw new BadGatewayError(response);
            case 503:
                throw new ServiceUnavailableError(response);
            case 504:
                throw new GatewayTimeoutError(response);
            default:
                if (response.status() >= 400 && response.status() < 500) {
                    throw new ClientError(response);
                } else if (response.status() >= 500 && response.status() < 600) {
                    throw new ServerError(response);
                } else {
                    throw new UnknownStatusError(response);
                }
        }
    }
}
